src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+        // and we WANT other tasks to go back to sleep if a task is already trying to create a new connection for this key.
+        //
+        // So the downsides are actually useful (we WANT task contention -
+        // to block other parallel tasks from trying to spawn a connection for this key if we are already doing so)
+        let inner_conn_lock = {
+            trace!("fallback attempt - getting keyed lock out of lockmap");
+            let guard = self.state.established_conn_writelock.guard();
+
+            let exist_conn_lock = self
+                .state
+                .established_conn_writelock
+                .get(&hash_key, &guard)
+                .unwrap();
+            trace!("fallback attempt - got keyed lock out of lockmap");
+            exist_conn_lock.as_ref().unwrap().clone()
+        };
+
+        debug!(
+            "appears we need a new conn, attempting to win connlock for wl key {:#?}",
+            workload_key
+        );
+        let got_conn = match inner_conn_lock.try_lock() {
+            Ok(_guard) => {
+                // BEGIN take inner writelock
+                // If we get here, it means the following are true:
+                // 1. We did not get a connection for our key.
+                // 2. We have the exclusive inner writelock to create a new connection for our key.
+                //
+                // So, carry on doing that.
+                debug!("nothing else is creating a conn and we won the lock, make one");
+                let pool_conn = self.spawn_new_pool_conn(workload_key.clone()).await;
+                let client = ConnClient {
+                    sender: pool_conn?,
+                    stream_count: Arc::new(AtomicU16::new(0)),
+                    stream_count_max: self.max_streamcount,
+                    wl_key: workload_key.clone(),
+                };
+
+                debug!(
+                    "starting new conn for key {:#?} with pk {:#?}",
+                    workload_key, pool_key
+                );
+                Some(client)
+                // END take inner writelock
+            }
+            Err(_) => {
+                debug!("we didnt' win the lock, something else is creating a conn, wait for it");
+                // If we get here, it means the following are true:
+                // 1. At one point, there was a preexisting conn in the pool for this key.
+                // 2. When we checked, we got nothing for that key.
+                // 3. We could not get the exclusive inner writelock to add a new one for this key.
+                // 4. Someone else got the exclusive inner writelock, and is adding a new one for this key.
+                //
+                // So, loop and wait for the pool_watcher to tell us a new conn was enpooled,
+                // so we can pull it out and check it.
+                loop {
+                    match self.pool_watcher.changed().await {
```

Comment:
hm.. Should the watchers be per-key? Seems like we are going to wake up a lot of things unnecessarily.

Reply:
Yes, but only if there's a lot of backpressure in the first place, so probably not a common problem.

I considered doing it per-key, but it complicates things considerably (we have another parallel keyed map, we need to synchronize it, etc etc)